### PR TITLE
Audit refactors - staking reward curve

### PR DESCRIPTION
**Pull request by hoani**
This PR primarily addresses concerns raised during an audit of the log2 function in the staking reward curve module.

## Addresses

- Add header documentation to the log2 function
- Add body documentation to the log2 function
- Taylor series term calculation should be its own function
- Misleading signed-ness of this function
- .pow(1) operation is redundant.
- Provide more thorough tests including panic tests and boundary cases
- Correct documentation links

## Concerns:

- None - it feels good to have the boundary tests implemented!

### CODE REVIEW

File: crml/staking/reward-curve/src/log.rs [hunk 1]
```diff
@@ -17,50 +17,89 @@
...
+		res /= 2 * k as u128 + 1;
+
+		res.try_into().unwrap()
+	}
+}
+
+/// Simple u32 power of 2 function - simply uses a bit shift
+fn pow2(n: u32) -> u32 {
+	return 1_u32 << n;
+}
```
Comment (cowboy-bebug) [inline crml/staking/reward-curve/src/log.rs:40]: 👍

File: crml/staking/reward-curve/src/log.rs [hunk 2]
```diff
@@ -17,50 +17,89 @@
...
 pub fn log2(p: u32, q: u32) -> u32 {
-	assert!(p >= q);
+	assert!(p >= q); // keep q/p bound to (0, 1]
 	assert!(p <= u32::max_value() / 2);
 
 	// This restriction should not be mandatory. But function is only tested and used for this.
 	assert!(p <= 1_000_000);
 	assert!(q <= 1_000_000);
 
+	// log2(1) = 0
```
Comment (cowboy-bebug) [inline crml/staking/reward-curve/src/log.rs:64]: 👍

File: crml/staking/reward-curve/src/log.rs [hunk 3]
```diff
@@ -17,50 +17,89 @@
...
+/// * result represents a per-million output of log2
+///   note: because result is u32, and the output of any log function
+///         in interval (0, 1) is negative, the output represents the
+///         absolute per million value of log2 and should be treated as
+///         a negative number
+///   note: yes, rational fractions are usually denoted p/q, but this
+///         function chooses to go q/p
 pub fn log2(p: u32, q: u32) -> u32 {
-	assert!(p >= q);
+	assert!(p >= q); // keep q/p bound to (0, 1]
```
Comment (cowboy-bebug) [inline crml/staking/reward-curve/src/log.rs:56]: Nice it makes sense now!

**Review by cowboy-bebug — APPROVED** [review]
Thanks for the refactor and helpful comments. This part of the code is definitely much more readable now than before 👍

File: crml/staking/reward-curve/src/log.rs [hunk 4]
```diff
@@ -17,50 +17,71 @@
...
+/// Simple u32 power of 2 function - simply uses a bit shift
+fn pow2(n: u32) -> u32 {
+	return 1_u32 << n;
+}
+
+/// Performs a log2 operation using a rational fraction
+///
+/// result = log2(q/p) where q/p is bound to (0, 1]
+/// Where:
+/// * p represents the numerator of the rational fraction input
```
Comment (cowboy-bebug) [inline crml/staking/reward-curve/src/log.rs]: Is `p` the denominator?

Comment (hoani) [reply crml/staking/reward-curve/src/log.rs]: Nice catch! will update this 👍

**Review by aliXsed — APPROVED** [review]
